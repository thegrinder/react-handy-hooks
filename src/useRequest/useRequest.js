import { useReducer, useMemo } from 'react';

const initialRequestState = {
  loading: false,
  initialLoad: true,
  succeeded: false,
  error: null,
  data: {},
};

const namespace = 'useRequest';
const actionTypes = {
  LOADING: `${namespace}/LOADING`,
  SUCCEEDED: `${namespace}/SUCCEEDED`,
  FAILED: `${namespace}/FAILED`,
  FULFILLED: `${namespace}/FULFILLED`,
};

const actionCreators = {
  loading: () => ({ type: actionTypes.LOADING }),
  succeeded: (data = {}) => ({ data, type: actionTypes.SUCCEEDED }),
  failed: (error) => ({ error, type: actionTypes.FAILED }),
  fulfilled: () => ({ type: actionTypes.FULFILLED }),
};

const reducer = (state, action) => {
  switch (action.type) {
    case actionTypes.LOADING:
      return {
        ...state,
        error: null,
        loading: true,
        succeeded: false,
      };
    case actionTypes.SUCCEEDED:
      return {
        ...state,
        loading: false,
        succeeded: true,
        initialLoad: false,
        data: action.data,
      };
    case actionTypes.FAILED:
      return {
        ...state,
        loading: false,
        initialLoad: false,
        error: action.error,
      };
    case actionTypes.FULFILLED:
      return {
        ...state,
        loading: false,
        initialLoad: false,
      };
    default:
      return state;
  }
};

const useRequest = () => {
  const [state, dispatch] = useReducer(reducer, initialRequestState);

  const boundActionCreators = useMemo(
    () => ({
      loading: () => dispatch(actionCreators.loading()),
      succeeded: (data = {}) => dispatch(actionCreators.succeeded(data)),
      failed: (error) => dispatch(actionCreators.failed(error)),
      fulfilled: () => dispatch(actionCreators.fulfilled()),
    }),
    []
  );

  const memoizedState = useMemo(() => state, [state]);

  return [memoizedState, boundActionCreators];
};

export default useRequest;
