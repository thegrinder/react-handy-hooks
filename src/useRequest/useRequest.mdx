---
name: useRequest
menu: hooks
---

# useRequest

useRequest is a hook that is designed to handle request state. It uses `useReducer` under the hood and returns state and three actions to update it.

## State

The return state has the following properties

- loading
- initialLoad
- succeeded
- error
- data

## Actions

### loading
- flips `loading` to `true`
- flips `succeeded` to `false`
- resets `error` to `null`

### succeeded
- flips `loading` to `false`
- flips `succeeded` to `true`
- flips `initialLoad` to `false`
- populates `data` with the given argument

### failed
- flips `loading` to `false`
- flips `initialLoad` to `false`
- populates `error` with the given argument

### fulfilled
- flips `loading` to `false`
- flips `initialLoad` to `false`

## Example

import { Playground } from 'docz';
import { useRequest } from '../';

<Playground>
  {() => {
    const UseRequestDemo = () => {
      const { state, actions } = useRequest();

      const mockRequest = () => new Promise((resolve, reject) => {
        return setTimeout(() => {
          Math.floor(Math.random() * 2) == 0
            ? resolve({ response: 'success' })
            : reject('oops!')
        }, 1000);
      });

      const makeRequest = () => {
        actions.loading();
        mockRequest()
          .then(response => actions.succeeded(response))
          .catch(error => actions.failed(error));
      };

      return (
        <div>
          <button onClick={makeRequest}>
            make request
          </button>
          {state.loading && (
            <div>loading...</div>
          )}
          {Object.keys(state.data).length !== 0 && (
            <div>
              response: {state.data.response}
            </div>
          )}
          {state.error && (
            <div>error: {state.error}</div>
          )}
        </div>
      )
    }

    return <UseRequestDemo />
  }}
</Playground>