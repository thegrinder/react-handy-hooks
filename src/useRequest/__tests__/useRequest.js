import { renderHook, act } from '@testing-library/react-hooks';
import useRequest from '../useRequest';

const initialRequestState = {
  loading: false,
  initialLoad: true,
  succeeded: false,
  error: null,
  data: {},
};

describe('useRequest', () => {
  it('should return the correct array', () => {
    const { result } = renderHook(() => useRequest());
    expect(result.current).toMatchSnapshot();
  });

  it('should handle loading action', () => {
    const { result } = renderHook(() => useRequest());
    act(() => {
      result.current[1].loading();
    });
    expect(result.current[0]).toEqual({
      ...initialRequestState,
      error: null,
      loading: true,
      succeeded: false,
    });
  });

  it('should handle succeeded action', () => {
    const { result } = renderHook(() => useRequest());
    const data = {};
    act(() => {
      result.current[1].succeeded(data);
    });
    expect(result.current[0]).toEqual({
      ...initialRequestState,
      data,
      loading: false,
      succeeded: true,
      initialLoad: false,
    });
  });

  it('should handle failed action', () => {
    const { result } = renderHook(() => useRequest());
    const error = 'error';
    act(() => {
      result.current[1].failed(error);
    });
    expect(result.current[0]).toEqual({
      ...initialRequestState,
      error,
      loading: false,
      initialLoad: false,
    });
  });

  it('should handle fulfilled action', () => {
    const { result } = renderHook(() => useRequest());
    act(() => {
      result.current[1].fulfilled();
    });
    expect(result.current[0]).toEqual({
      ...initialRequestState,
      loading: false,
      initialLoad: false,
    });
  });
});
