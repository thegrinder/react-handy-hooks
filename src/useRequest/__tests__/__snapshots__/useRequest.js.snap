// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useRequest should return the correct array 1`] = `
Array [
  Object {
    "data": Object {},
    "error": null,
    "initialLoad": true,
    "loading": false,
    "succeeded": false,
  },
  Object {
    "failed": [Function],
    "fulfilled": [Function],
    "loading": [Function],
    "succeeded": [Function],
  },
]
`;
